test/unit/matrix/test_copy.cpp

```diff
@@ -0,0 +1,365 @@
+      Matrix<const TypeParam, Device::CPU> mat_src_const = std::move(mat_src);
+
+      memory::MemoryView<TypeParam, Device::GPU> mem_gpu1(layout.minMemSize());
+      Matrix<TypeParam, Device::GPU> mat_gpu1(distribution, layout, mem_gpu1());
+
+      memory::MemoryView<TypeParam, Device::GPU> mem_gpu2(layout.minMemSize());
+      Matrix<TypeParam, Device::GPU> mat_gpu2(distribution, layout, mem_gpu2());
+
+      memory::MemoryView<TypeParam, Device::CPU> mem_dst(layout.minMemSize());
+      Matrix<TypeParam, Device::CPU> mat_dst(distribution, layout, mem_dst());
+      set(mat_dst, outputValues<TypeParam>);
+
+      copy(mat_src_const, mat_gpu1);
+      copy(mat_gpu1, mat_gpu2);
+      copy(mat_gpu2, mat_dst);
+
+      CHECK_MATRIX_NEAR(inputValues<TypeParam>, mat_dst, 0, TypeUtilities<TypeParam>::error);
+    }
+  }
+}
+#endif
+
+struct SubMatrixCopyConfig {
+  GlobalElementSize full_in;
+  GlobalElementSize full_out;
+
+  TileElementSize tile_size;
+
+  GlobalElementIndex sub_origin_in;
+  GlobalElementIndex sub_origin_out;
+
+  GlobalElementSize sub_size;
+};
+
+comm::Index2D alignSubRankIndex(const Distribution& dist_in, const GlobalElementIndex& offset_in,
+                                const GlobalElementIndex& offset_out) {
+  const auto pos_mod = [](const auto& a, const auto& b) {
+    const auto mod = a % b;
+    return (mod >= 0) ? mod : (mod + b);
+  };
+
+  const comm::Size2D grid_size = dist_in.grid_size();
+  const comm::Index2D sub_rank(dist_in.rank_global_element<Coord::Row>(offset_in.row()),
+                               dist_in.rank_global_element<Coord::Col>(offset_in.col()));
+  const GlobalTileIndex offset_rank(offset_out.row() / dist_in.block_size().rows(),
+                                    offset_out.col() / dist_in.block_size().cols());
+
+  return {pos_mod(sub_rank.row() - static_cast<comm::IndexT_MPI>(offset_rank.row()), grid_size.rows()),
+          pos_mod(sub_rank.col() - static_cast<comm::IndexT_MPI>(offset_rank.col()), grid_size.cols())};
+}
+
+bool isFullMatrix(const Distribution& dist_full, const GlobalElementIndex& sub_origin,
+                  const GlobalElementSize& sub_size) noexcept {
+  return sub_origin == GlobalElementIndex{0, 0} && sub_size == dist_full.size();
+}
+
+template <class ElementGetter>
+auto subValues(ElementGetter&& fullValues, const GlobalElementIndex& offset) {
+  return [fullValues, offset = sizeFromOrigin(offset)](const GlobalElementIndex& ij) {
+    return fullValues(ij + offset);
+  };
+}
+
+template <class OutsideElementGetter, class InsideElementGetter>
+auto mixValues(const GlobalElementIndex& offset, const GlobalElementSize& sub_size,
+               InsideElementGetter&& insideValues, OutsideElementGetter&& outsideValues) {
+  return [outsideValues, insideValues, offset, sub_size](const GlobalElementIndex& ij) {
+    if (ij.isInSub(offset, sub_size))
+      return insideValues(ij - common::sizeFromOrigin(offset));
+    else
+      return outsideValues(ij);
+  };
+}
```

Comment:
These helpers might be useful elsewhere, and I will probably move them in some "common" place so that they can be used in other tests too. (Probably this will happen in #969)